Installation troubles...

I've wanted to try out arrayfire to speed up some calculations (in rust, not cpp), but after installing it to `/opt/arrayfire` and setting the environmental variables: 

```
$ echo $AF_PATH
/opt/arrayfire
```

and 

```
$ echo $LD_LIBRARY_PATH
/opt/arrayfire/lib64 // or /opt/arrayfire/lib, I tried both, and even created two duplicates of the directory: lib and lib64
```

but an error occurs when trying out something simple: 


```
use arrayfire as af;
use af::{Array, Dim4, randu, print};

fn main() {
    let dims = Dim4::new(&[5, 3, 1, 1]);
    let a = randu::<f32>(dims);
    print(&a);
}
```

namely

```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-m64" "-L" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.1aoax2tfw9s7f78r.rcgu.o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.1ebkme1zvs2rtcr7.rcgu.o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.1gmaanxh4i9cz6qp.rcgu.o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.1latvcgg1hjepj0c.rcgu.o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.2h3s8ekgajz7g7qh.rcgu.o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.3v7ev815qfdp4atk.rcgu.o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.4pu8k9rtk6cox549.rcgu.o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.51rqsp4z7bhiengm.rcgu.o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.7f37ybetqa4m08e.rcgu.o" "-o" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/playground.obgl6p0h9sv65fp.rcgu.o" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps" "-L" "/opt/arrayfire/lib" "-L" "/opt/arrayfire/lib64" "-L" "/usr/local/lib" "-L" "/usr/lib" "-L" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/build/glfw-sys-18e0e4baf696aed1/out/lib" "-L" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libarrayfire-b27552ba11a4042b.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libhalf-bd08fd2e2cef5202.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/liblibc-46922849d95b2796.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libnum-7f27e0ad131389b6.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libnum_rational-ed1f558895f2b299.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libnum_iter-2202a92a46fcf5e9.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libnum_complex-0fec95524a779422.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libnum_bigint-a949f801f8a854f4.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libnum_integer-f55caa9968b5aade.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/libnum_traits-4efa7eecee5ac88c.rlib" "/Users/pvdklei/Desktop/Code/graphics/playground/target/debug/deps/liblazy_static-c99c60063522767f.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libstd-f13f993217d0ff19.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-fd322230517448c5.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libhashbrown-bfac32b2346e6f4f.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_alloc-366aa1278b6a1212.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libbacktrace-e3336f6b916a3e06.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libbacktrace_sys-f90e2fda5c4354a8.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_demangle-3f082039d6523013.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libunwind-0e0d502cf4053e50.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcfg_if-a69411463c63ffac.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liblibc-63b216fe2d62ff77.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liballoc-4e60e72e5eb2e571.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_std_workspace_core-4c26a78942ac2ec8.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcore-81342cbee8440a1f.rlib" "/Users/pvdklei/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-f0244eb4ede32fb8.rlib" "-lSystem" "-lresolv" "-lc" "-lm"
  = note: Undefined symbols for architecture x86_64:
            "_af_get_last_error", referenced from:
                arrayfire::core::error::get_last_error::hc6a8d9d99590c587 in libarrayfire-b27552ba11a4042b.rlib(arrayfire-b27552ba11a4042b.arrayfire.8784b43a-cgu.11.rcgu.o)
            "_af_free_host", referenced from:
                arrayfire::core::util::free_host::h69be059379ca13ca in libarrayfire-b27552ba11a4042b.rlib(arrayfire-b27552ba11a4042b.arrayfire.8784b43a-cgu.10.rcgu.o)
            "_af_release_array", referenced from:
                _$LT$arrayfire..core..array..Array$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::he9ec9b77d35bedf3 in libarrayfire-b27552ba11a4042b.rlib(arrayfire-b27552ba11a4042b.arrayfire.8784b43a-cgu.4.rcgu.o)
            "_af_print_array_gen", referenced from:
                arrayfire::core::array::print::h131ea6d8a0528025 in playground.3v7ev815qfdp4atk.rcgu.o
            "_af_randu", referenced from:
                arrayfire::core::random::randu::hb78865a4b22d2eac in playground.1aoax2tfw9s7f78r.rcgu.o
          ld: symbol(s) not found for architecture x86_64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: aborting due to previous error; 1 warning emitted
```

How can I solve this?

--- a/README.md
+++ b/README.md
@@ -82,6 +82,14 @@ Create a 5-by-3 matrix of random floats on the GPU
 ...
 ```
 
+### Troubleshooting
+
+If the build command fails with undefined references errors even after taking care of environment
+variables, we recommend doing a `cargo clean` and re-running `cargo build` or `cargo test`.
+
+You can also use some environment variables mentioned in our [book][23], such as `AF_PRINT_ERRORS`
+to print more elaborate error messages to console.
+
 ## Acknowledgements
 
 The ArrayFire library is written by developers at [ArrayFire][14] LLC with [contributions][15]
@@ -118,3 +126,4 @@ the DARPA SBIR Program Office.
 [20]: https://img.shields.io/badge/arrayfire-google--groups-orange
 [21]: http://arrayfire.org/arrayfire-rust/book/index.html
 [22]: https://img.shields.io/badge/arrayfire-mdbook-073763?logo=readthedocs
+[23]: http://arrayfire.org/arrayfire-rust/book/configuring_arrayfire_environment.html

--- a/src/core/macros.rs
+++ b/src/core/macros.rs
@@ -298,6 +298,25 @@ macro_rules! eval {
     };
 }
 
+/// Create an array of given shape filled with a single value a.k.a constant array
+///
+/// # Examples
+///
+/// ```rust
+/// # use arrayfire::{constant};
+/// let _zeros_1d = constant!(0.0f32; 10);
+/// let _ones_3d = constant!(1u32; 3, 3, 3);
+///
+/// let dim = 10;
+/// let mix_shape = constant!(42.0f32; dim, 10);
+/// ```
+#[macro_export]
+macro_rules! constant {
+    ($value:expr; $($dim:expr),+) => {
+        $crate::constant($value, $crate::dim4!($($dim),*))
+    };
+}
+
 #[cfg(test)]
 mod tests {
     use super::super::array::Array;
@@ -424,5 +443,16 @@ mod tests {
 
         eval!(a[indices, seq4gen] = b);
         // ANCHOR_END: macro_seq_array_assign
+    }
+
+    #[test]
+    fn constant_macro() {
+        let _zeros_1d = constant!(0.0f32; 10);
+        let _zeros_2d = constant!(0.0f64; 5, 5);
+        let _ones_3d = constant!(1u32; 3, 3, 3);
+        let _twos_4d = constant!(2u16; 2, 2, 2, 2);
+
+        let dim = 10;
+        let _mix_shape = constant!(42.0f32; dim, 10);
     }
 }

--- a/src/core/macros.rs
+++ b/src/core/macros.rs
@@ -317,6 +317,38 @@ macro_rules! constant {
     };
 }
 
+/// Create an array of given shape sampled from uniform distribution
+///
+/// If no type argument is specified, the data type defaults to 32 bit floats.
+///
+/// # Examples
+///
+/// ```rust
+/// # use arrayfire::{randu};
+/// let mat10x10 = randu!(10, 10);
+/// ```
+#[macro_export]
+macro_rules! randu {
+    ($($dim:expr),+) => { $crate::randu::<f32>($crate::dim4!($($dim),*)) };
+    ($type:ty; $($dim:expr),+) => { $crate::randu::<$type>($crate::dim4!($($dim),*)) };
+}
+
+/// Create an array of given shape sampled from normal distribution
+///
+/// If no type argument is specified, the data type defaults to 32 bit floats.
+///
+/// # Examples
+///
+/// ```rust
+/// # use arrayfire::{randn};
+/// let mat10x10 = randn!(10, 10);
+/// ```
+#[macro_export]
+macro_rules! randn {
+    ($($dim:expr),+) => { $crate::randn::<f32>($crate::dim4!($($dim),*)) };
+    ($type:ty; $($dim:expr),+) => { $crate::randn::<$type>($crate::dim4!($($dim),*)) };
+}
+
 #[cfg(test)]
 mod tests {
     use super::super::array::Array;
@@ -454,5 +486,13 @@ mod tests {
 
         let dim = 10;
         let _mix_shape = constant!(42.0f32; dim, 10);
+    }
+
+    #[test]
+    fn rand_macro() {
+        let _ru5x5 = randu!(5, 5);
+        let _rn5x5 = randn!(5, 5);
+        let _ruu32_5x5 = randu!(u32; 5, 5);
+        let _ruu8_5x5 = randu!(u8; 5, 5);
     }
 }

